Bluetooth: controller: Introduce ULL LLL architecture

cvinayak opened the pull request:
This Pull Request introduces a new, improved Link Layer based on the concept of split responsibilities:

* The Upper Link Layer (ULL) is in charge of control procedures, inter-event scheduling and overal role management. The code for the ULL is shared among all hardware implementations.
* The Lower Link Layer (LLL) is responsible for the intra-event scheduling and radio access.

The communication between ULL and LLL is achieved through a set of FIFOs that contain both control and data packets.

cvinayak committed:
Bluetooth: controller: Refactor the internal LL interfaces

Refactored the internal LL interfaces to have return value
to match the HCI error code u8_t data type.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Introduce ULL LLL architecture

This is a squash merge of commits introducing the new split
Upper Link Layer and Lower Link Layer architecture of the
Bluetooth Low Energy controller.

This introduces a new, improved Link Layer based on the
concept of split responsibilities; The Upper Link Layer
(ULL) is in charge of control procedures, inter-event
scheduling and overall role management. The code for the
ULL is shared among all hardware implementations. The
Lower Link Layer (LLL) is responsible for the intra-event
scheduling and vendor specific radio hardware access.

The communication between ULL and LLL is achieved through
a set of FIFOs that contain both control and data packets.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
Signed-off-by: Alberto Escolar Piedras <alpi@oticon.com>
Signed-off-by: Wolfgang Puffitsch <wopu@oticon.com>
Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak committed:
Bluetooth: controller: Preliminary work in progress mesh ext.

Preliminary work done towards Mesh extensions on the old LL
architecture implementation.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Updates to prep to introduce new arch

Missing updates to old architecture implementation towards
introduction of new ULL LLL architecture.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: shell: Updates related to new ULL LLL architecture

Updates related to new ULL LLL controller architecture.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_tmp.c:
```diff
@@ -0,0 +1,240 @@
+	struct lll_prepare_param *p = param;
+	int err;
+
+	printk("\t\tlll_tmp_prepare (%p) enter.\n", p->param);
```
Are those `printk` staatements intentional?

cvinayak replied:
tmp was  a template role, this I will remove as part of the rework, this should not be part of the merge into master.

carlescufi commented on subsys/bluetooth/controller/util/mfifo.h:
```diff
@@ -0,0 +1,177 @@
+			u8_t const n; /* TODO: const, optimise RAM use */ \
+			u8_t f; \
+			u8_t l; \
+			u8_t m[sz * ((cnt) + 1)]; \
```
If you are using this only to store ring buffers of pointers, then you could simplify by:
`void* m[(cnt) + 1];` and get rid of `s` and `sz` altogether.

carlescufi replied:
This would also get rid of all the multiplications in the getters/setters

cvinayak replied:
I will take relook at this implementation once I resolve all the checkpatch issues.

carlescufi commented on subsys/bluetooth/controller/util/mfifo.h:
```diff
@@ -0,0 +1,177 @@
+{
+	void **p = (void **)(fifo + (*last) * size);
+
+	*p = mem;
```
If you need to support non-pointer mfifos then this should be a `memcpy`

cvinayak replied:
Agree.

carlescufi commented on subsys/bluetooth/controller/util/mfifo.h:
```diff
@@ -0,0 +1,177 @@
+
```
This file needs comments on each function explaining what they do.

mped-oticon replied:
I've documented this in https://github.com/zephyrproject-rtos/zephyr/pull/12167

mped-oticon replied:
Note that  #12167 was spun out of https://github.com/NordicPlayground/zephyr-bt/pull/26 as that was the intersection between non-split and split stack.

carlescufi replied:
Thank you @mped-oticon, that is a great effort. A while back I added the `memq_link_t` in order to make the code more readable, and with your comments this closes the circle.

carlescufi replied:
>I've documented this in #12167

That PR doesn't include `mfifo` but you have another branch for that.

Vudentz replied:
Remind me why we need a different fifo/list implementation?

carlescufi replied:
For 3 reasons:

1) Memory efficiency and code size
2) Latency: this is lockless unlike the kernel's FIFOs, never disables interrupts
3) Independence from the kernel. This is only used on interrupt context, should never block or use kernel primitives

carlescufi commented on subsys/bluetooth/controller/util/mfifo.h:
```diff
@@ -0,0 +1,177 @@
+		return NULL;
+	}
+
+	return *((void **)(fifo + first * size));
```
This will only work with `void*` MFIFOs

mped-oticon replied:
Correct. I've written about this in the documentation. See prev comment.

carlescufi replied:
I guess you mean next comment, since the previous PR doesn't include `mfifo`

carlescufi commented on subsys/bluetooth/controller/util/mfifo.h:
```diff
@@ -0,0 +1,177 @@
+		return NULL;
+	}
+
+	mem = *((void **)(fifo + _first * size));
```
This will only work with `void*` MFIFOs

mped-oticon replied:
See header of https://github.com/mped-oticon/zephyr-bt/blob/783481527a1c435251484783bd4ef01bd237654b/subsys/bluetooth/controller/util/mfifo.h

carlescufi replied:
Oh nice! Will you submit this after we merge this PR?

carlescufi commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -0,0 +1,4730 @@
+	}
+
+	if (!cmd) {
+		if (conn->llcp_req != conn->llcp_ack) {
```
The problem I see with these is that the register are typically loaded first, and then the comparison is executed:

```
                if (conn->llcp_req != conn->llcp_ack) {
   13916:       f890 30d0       ldrb.w  r3, [r0, #208]  ; 0xd0
   1391a:       f890 20d1       ldrb.w  r2, [r0, #209]  ; 0xd1
   1391e:       429a            cmp     r2, r3
```

Which means that if you get interrupted between `1391a` and `1391e` you will be comparing a stale value.

cvinayak replied:
I will add a commit to this PR to handle context-safety of the control procedures.

cvinayak committed:
Bluetooth: controller: Fix control procedure context safety

Fix the control procedure context safety by adding checks in
thread mode control path to detect pre-emption by interrupt.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

carlescufi commented on subsys/bluetooth/controller/ll_sw/ull_conn.c:
```diff
@@ -331,6 +337,12 @@ u8_t ll_feature_req_send(u16_t handle)
 		return BT_HCI_ERR_CMD_DISALLOWED;
 	}
 
+	conn->llcp_req++;
```
On Armv7-M this would be better served by using `atomic_cas()` is my opinion. I am not sure about other architectures (ping @mped-oticon and @aescolar) but on Armv7-M `atomic_cas()` uses `CONFIG_ATOMIC_OPERATIONS_BUILTIN` which then translates it into the following, which in my opinion is going to be smaller in size than your code once assembled.
Note that on Armv6-M `atomic_cas()` disables interrupts and so there you will need the software implementation you have here.

```
        return __atomic_compare_exchange_n(target, &old_value, new_value,
     4d4:       4b17            ldr     r3, [pc, #92]   ; (534 <bt_ready+0x74>)
     4d6:       9403            str     r4, [sp, #12]
     4d8:       2104            movs    r1, #4
     4da:       f3bf 8f5b       dmb     ish
     4de:       e853 2f00       ldrex   r2, [r3]
     4e2:       42a2            cmp     r2, r4
     4e4:       d103            bne.n   4ee <bt_ready+0x2e>
     4e6:       e843 1000       strex   r0, r1, [r3]
     4ea:       2800            cmp     r0, #0
     4ec:       d1f7            bne.n   4de <bt_ready+0x1e>
```

mped-oticon replied:
Note: Oticon's mcu does not have atomic_cas.

koffes replied:
Since the value is incremented, should we use `atomic_inc(&conn->llcp_req)` ?

mped-oticon reviewed (approved):
Remove DNM?

thoh-ot reviewed (approved):
QACK

jhedberg commented on subsys/bluetooth/shell/hci.h:
```diff
@@ -0,0 +1,12 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __HCI_H_
```
All header guards in Zephyr were recently changed to be constructed out of the full path. This file's should then be `ZEPHYR_SUBSYS_BLUETOOTH_SHELL_HCI_H_`

jhedberg replied:
Actually, does this file even need a guard? Isn't it completely internal?

cvinayak replied:
Agree will remove it.

jhedberg commented on subsys/bluetooth/shell/ll.c:
```diff
@@ -258,9 +257,22 @@ int cmd_advx(const struct shell *shell, size_t argc, char *argv[])
 	}
 
+#if defined(CONFIG_BT_LL_SW_SPLIT)
+	printk("ad data set...");
```
Is it appropriate for a shell module to use printk? Isn't the recommendation to use the shell's own output APIs?

cvinayak replied:
This is an oversight, will fix it.

Vudentz commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -463,6 +469,12 @@ u8_t ll_start_enc_req_send(u16_t handle, u8_t error_code,
 			return BT_HCI_ERR_CMD_DISALLOWED;
 		}
 
+		conn->llcp_req++;
```
There is some many duplicates of this logic bellow, can't we make a helper function to check llcp_req against llcp_ack?

cvinayak replied:
Agree. Can we agree to create an issue and assign it to me? This is mentioned by @carlescufi to evaluate the use of atomic_CAS instead.

Vudentz replied:
No problem, lets have it separately with use of atomic as well.

mped-oticon replied:
> Agree. Can we agree to create an issue and assign it to me? This is mentioned by @carlescufi to evaluate the use of atomic_CAS instead.

Note: Oticon's mcu does not have atomic_cas.
Otherwise, fully agree about helper function or macro.

jhedberg commented on subsys/bluetooth/shell/ll.c:
```diff
@@ -330,3 +342,67 @@ int cmd_scanx(const struct shell *shell, size_t  argc, char *argv[])
+		err = ull_tmp_disable(handle);
+	}
+
+	shell_print(shell, "Done (%d).\n", err);
```
shell_print() should not have a line terminator, so the \n needs to be removed

jhedberg commented on subsys/bluetooth/shell/ll.c:
```diff
@@ -330,3 +342,67 @@ int cmd_scanx(const struct shell *shell, size_t  argc, char *argv[])
+
+	err = ull_tmp_data_send(handle, 0, NULL);
+
+	shell_print(shell, "Done (%d).\n", err);
```
same here
